### DISCUSSION TITLE
Is there an expected timeline for the stable release of 17.0?

### DISCUSSION THREAD

**Discussion by bsmith-self** [graphql/graphql-js#3892] [General]
We're waiting to upgrade from Node 14 because we don't want to go to an alpha version of `graphql` for Node 18 support, but trying to decide whether we need to go ahead and upgrade to 16 and then go to 18 a few months from now when 16 is EOL, or just wait and go straight to 18 (if the stable `graphql` release is right around the corner)

I realize these things can be uncertain, but any information about the time-scale (days? weeks? months?) would be appreciated

**Comment by torbjornvatn** [comment]
@IvanGoncharov any updates on this!

We really need https://github.com/graphql/graphql-js/pull/2757 _Support returning async iterables from resolver functions_ in some of our production environments, but we have to keep building a separate version with the alpha version because we don't want to use it in all our environments

**Reply by philippemiguet** [reply]
Same, we really need this patch in production for async iterables. We have our event loop blocked by the `execute` method when a resolver returns a large array of results and that method needs to generate promises of all field resolvers.

**Comment by leszekhanusz** [comment]
See also [graphql-python/gql#427](https://github.com/graphql-python/gql/issues/427).

Is there anything blocking a stable release of 17.0?

**Comment by jarlefosen** [comment]
Hi, is there a roadmap for what's needed in order to release v17? cc @IvanGoncharov

**Comment by ephemer** [comment]
I'm curious why 17.x has been "stuck" at alpha for two years. Is it an indication that the stability is not where it should be on certain features? Is there something the community can support with?

We would love to use `@stream` and/or `@defer` (we can model our data either way – we just need delayed responses of some kind) but it's unclear to us what is blocking these features from reaching the mainstream, or even a `beta` channel. `alpha` sounds very unstable but I'm not sure it truly is.

@JoviDeCroock would you be able to say a few words about the decision-making here? Is some part of this feature set very unstable and others relatively stable? Is there some asynchronous response feature that could be backported to 16.x instead? Or how can we best interpret the current situation? Thank you!

**Reply by JoviDeCroock** [reply]
That's a very good question. I want to start out by prefacing this with, both stream as well as defer are still in the spec traversal process so even if we'd release v17 this would not be under the normal `execute` path but under an `executeIncrementally` export that highlights that this spec is in evolution. With [GraphQL Yoga](https://the-guild.dev/graphql/yoga-server/docs/features/defer-stream) you can already try out these incremental delivery directives today.

When it comes to v17, let me say that I've only joined this repository recently as a maintainer and I'm very much still struggling to get a picture of what has happened in v17, see my work in https://github.com/graphql/graphql-js/pull/4310 to see what I've been up to. I've also taken a step back from v17 because the first and most important thing I've noticed is that people have been stuck at previous major versions i.e. v15 an v14 still have a lot of downloads. My current assumption is that the lack of documentation and upgrade guides attribute to this, hence why my top priority has been to get [a new docs site out](https://www.graphql-js.org/).

Now, I won't claim that v17 is unstable in the slightest, we want to get this right. We want to get a good mental picture of what has changed in v17 and then evaluate whether this is what we want. There are also still some open things we need to tackle, as can be seen in https://github.com/graphql/graphql-js/issues/4205. One last thing is that defer/stream has been changing its response format a few times, and as part of releasing v17 we should not break that, we want to ease this upgrade journey which will need us to support the last 2 formats so i.e. apollo-server/GraphQL-Yoga can upgrade their consumers without needing userland solutions.

I know that most of this isn't fully answering your question for a timeline, forgive me for that. I want to move everything along but getting enough context to make decisions will take time.

**Reply by JoviDeCroock** [reply]
I'd love to also CC others in this thread and apologize for not getting a response to all of this earlier

@torbjornvatn @jarlefosen @leszekhanusz

**Reply by benjie** [reply]
As Jovi says, `@stream` and `@defer` are still being developed; if you wish to join these efforts you may join the WG: https://github.com/graphql/defer-stream-wg/ . The latest spanner in the works is described here for example: https://github.com/graphql/defer-stream-wg/discussions/100 (though you may want to familiarize yourself with the previous work in order to understand the complexity of this specific issue).

We are also in the process of getting the agreed parts of stream and defer through the spec review process, including bringing small parts for review to each main GraphQL WG and splitting the work into smaller more manageable chunks. Incremental delivery is a very large feature, with a lot of potential pitfalls if implemented naively (see the response explosion attacks easily possible in the pre-June 2023 response format, for example); coming up with a solution that works for all clients, isn't too complex, but doesn't suffer from these issues (or at least not as much) is a significant undertaking - and the people working on it are all volunteers.

**Reply by Urigo** [reply]
I would add to that that one of the best things you can do to support the work is to start using it already (with extreme caution) and provide feedback from your real use case in production to the WG
If you need help using these today feel free to reach out

**Reply by ephemer** [reply]
@Urigo @benjie @JoviDeCroock thank you kindly for your efforts and replies here. It is heartening to hear that the work on these topics is ongoing. 

I don't think I'm the right person to join the WG: I'm too much of a pragmatist and uninvested in the deepest edge cases to be much help hammering out the final details. For example, my suggestion for the case that you shared @benjie would be to simply disallow that combination of effects (i.e. simply deem the query invalid). Since `@stream` and `@defer` don't yet exist anywhere officially I imagine that to be feasible. I believe the community would benefit a lot more from having a working (if ever so slightly limited) implementation today, rather than waiting to solve cases like that to perfection. Again, just my 2c – and my currency probably isn't worth much in the world of standards and specifications.

It is very helpful to hear that v17 alpha is not considered unstable, but rather that certain edge cases need to be ironed out, and adoption paths need to be clarified. FYI for us the only challenge updating from 15.x to 16.x was the amount of TypeScript errors the change caused – presumably Apollo Server also smoothened out some of the differences under the hood as well though.

If there is no specific reason not to use v17 in production, I would be happy to give it a shot and report back with any issues we may run into. FWIW we're using Apollo Server 4. We just started the new year and have bitten off probably more than we can chew already, but I'd be happy to introduce this change soon to give it a shot.

**Reply by benjie** [reply]
> For example, my suggestion for the case that you shared @benjie would be to simply disallow that combination of effects (i.e. simply deem the query invalid). 

That’s currently the leading proposal so it sounds like you’d fit right in!

**Reply by philippemiguet** [reply]
For what it's worth, we have been running v17-alpha in production for the past 6 months without a single problem on a service :)
